corehq/motech/repeaters/migrations/0016_add_indexes.py

```diff
@@ -0,0 +1,62 @@
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+    atomic = False
+
+    dependencies = [
+        ("repeaters", "0015_repeater_max_workers"),
+    ]
+
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            state_operations=[
+                migrations.AlterField(
+                    model_name="repeatrecord",
+                    name="state",
+                    field=models.PositiveSmallIntegerField(
+                        choices=[
+                            (1, "Pending"),
+                            (2, "Failed"),
+                            (4, "Succeeded"),
+                            (8, "Cancelled"),
+                            (16, "Empty"),
+                            (32, "Invalid Payload"),
+                        ],
+                        db_index=True,
+                        default=1,
+                    ),
+                ),
+                migrations.AddIndex(
+                    model_name="repeater",
+                    index=models.Index(
+                        condition=models.Q(("is_deleted", False), ("is_paused", False)),
+                        fields=["next_attempt_at"],
+                        name="next_attempt_at_partial_idx",
+                    ),
+                ),
+            ],
+
+            database_operations=[
+                migrations.RunSQL(
+                    sql="""
+                    CREATE INDEX CONCURRENTLY "repeaters_repeatrecord_state_8055083b"
+                    ON "repeaters_repeatrecord" ("state");
+                    """,
+                    reverse_sql="""
+                    DROP INDEX CONCURRENTLY "repeaters_repeatrecord_state_8055083b";
+                    """
+                ),
+                migrations.RunSQL(
+                    sql="""
+                    CREATE INDEX CONCURRENTLY "next_attempt_at_partial_idx"
+                    ON "repeaters_repeater" ("next_attempt_at")
+                    WHERE (NOT "is_deleted" AND NOT "is_paused");
```

Comment:
I analyzed the query used by `Repeater.objects.get_all_ready_ids_by_domain()`, and this is the result on Staging:
```
> EXPLAIN ANALYZE SELECT "repeaters_repeater"."domain", "repeaters_repeater"."id_"
  FROM "repeaters_repeater"
    INNER JOIN "repeaters_repeatrecord"
    ON ("repeaters_repeater"."id_" = "repeaters_repeatrecord"."repeater_id_")
  WHERE (
    NOT "repeaters_repeater"."is_deleted"
    AND NOT "repeaters_repeater"."is_paused"
    AND (
      "repeaters_repeater"."next_attempt_at" IS NULL
      OR "repeaters_repeater"."next_attempt_at" <= '2024-10-19 22:11:36.310082'
    )
    AND "repeaters_repeatrecord"."state" IN (1, 2)
  );

                                                                                      QUERY PLAN                                                                                      
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=8423.11..141317.33 rows=52115 width=30) (actual time=61.472..1547.920 rows=638862 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Hash Join  (cost=7423.11..135105.83 rows=21715 width=30) (actual time=56.214..1224.831 rows=212954 loops=3)
         Hash Cond: (repeaters_repeatrecord.repeater_id_ = repeaters_repeater.id_)
         ->  Parallel Bitmap Heap Scan on repeaters_repeatrecord  (cost=7289.92..134244.51 rows=276830 width=16) (actual time=52.249..1077.808 rows=223024 loops=3)
               Recheck Cond: (state = ANY ('{1,2}'::integer[]))
               Heap Blocks: exact=16852
               ->  Bitmap Index Scan on repeaters_repeatrecord_state_8055083b  (cost=0.00..7123.82 rows=664393 width=0) (actual time=44.255..44.256 rows=669452 loops=1)
                     Index Cond: (state = ANY ('{1,2}'::integer[]))
         ->  Hash  (cost=131.11..131.11 rows=167 width=30) (actual time=0.713..0.722 rows=163 loops=3)
               Buckets: 1024  Batches: 1  Memory Usage: 18kB
               ->  Bitmap Heap Scan on repeaters_repeater  (cost=5.84..131.11 rows=167 width=30) (actual time=0.089..0.660 rows=163 loops=3)
                     Filter: ((NOT is_deleted) AND (NOT is_paused) AND ((next_attempt_at IS NULL) OR (next_attempt_at <= '2024-10-19 22:11:36.310082+00'::timestamp with time zone)))
                     Rows Removed by Filter: 40
                     Heap Blocks: exact=67
                     ->  Bitmap Index Scan on repeaters_repeater_is_deleted_08441bf0  (cost=0.00..5.80 rows=203 width=0) (actual time=0.053..0.059 rows=253 loops=3)
                           Index Cond: (is_deleted = false)
 Planning Time: 0.874 ms
 Execution Time: 1623.954 ms
(20 rows)
```

The "next_attempt_at_partial_idx" was chosen to match the filter in that query plan, but Postgres is not using the index.

Should we drop it, change it, or leave it?

Reply:
How much data is present on staging? If not much then those two indexes may have had very similar outcomes and the choice of which to use may not have been very meaningful.

I'd check the same or a similar query on prod to see if you get the same query plan.

Do we need both indexes? Would it break other queries if we added the partial condition `NOT "is_deleted" AND NOT "is_paused"` on `repeaters_repeatrecord_state_8055083b` and eliminated the `next_attempt_at_partial_idx` index?

Reply:
Interesting! Thank you!

Reply:
Very nice. It uses the index.
```
QUERY PLAN
----------
 Gather  (cost=1078.24..298737.78 rows=91675 width=30) (actual time=1860.938..7291.760 rows=952533 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Parallel Append  (cost=78.24..288570.28 rows=91675 width=30) (actual time=1103.767..5783.887 rows=317511 loops=3)
         ->  Hash Join  (cost=132.62..143624.77 rows=31897 width=30) (actual time=1.539..2456.512 rows=317502 loops=3)
               Hash Cond: (repeaters_repeatrecord.repeater_id_ = repeaters_repeater.id_)
               ->  Parallel Seq Scan on repeaters_repeatrecord  (cost=0.00..142389.57 rows=419192 width=16) (actual time=1.080..2293.795 rows=327791 loops=3)
                     Filter: (state = ANY ('{1,2}'::integer[]))
                     Rows Removed by Filter: 1776517
               ->  Hash  (cost=130.60..130.60 rows=162 width=30) (actual time=0.357..0.359 rows=153 loops=3)
                     Buckets: 1024  Batches: 1  Memory Usage: 17kB
                     ->  Bitmap Heap Scan on repeaters_repeater  (cost=5.84..130.60 rows=162 width=30) (actual time=0.067..0.295 rows=153 loops=3)
                           Filter: ((NOT is_deleted) AND (NOT is_paused) AND (next_attempt_at IS NULL))
                           Rows Removed by Filter: 50
                           Heap Blocks: exact=67
                           ->  Bitmap Index Scan on repeaters_repeater_is_deleted_08441bf0  (cost=0.00..5.80 rows=203 width=0) (actual time=0.044..0.045 rows=253 loops=3)
                                 Index Cond: (is_deleted = false)
         ->  Hash Join  (cost=78.24..143570.39 rows=6301 width=30) (actual time=1655.248..4904.097 rows=14 loops=2)
               Hash Cond: (repeaters_repeatrecord_1.repeater_id_ = repeaters_repeater_1.id_)
               ->  Parallel Seq Scan on repeaters_repeatrecord repeaters_repeatrecord_1  (cost=0.00..142389.57 rows=419192 width=16) (actual time=0.043..4756.595 rows=491687 loops=2)
                     Filter: (state = ANY ('{1,2}'::integer[]))
                     Rows Removed by Filter: 2664776
               ->  Hash  (cost=77.84..77.84 rows=32 width=30) (actual time=39.663..39.664 rows=10 loops=2)
                     Buckets: 1024  Batches: 1  Memory Usage: 9kB
                     ->  Bitmap Heap Scan on repeaters_repeater repeaters_repeater_1  (cost=4.39..77.84 rows=32 width=30) (actual time=39.626..39.647 rows=10 loops=2)
                           Recheck Cond: ((next_attempt_at <= '2024-10-19 22:11:36.310082+00'::timestamp with time zone) AND (NOT is_deleted) AND (NOT is_paused))
                           Heap Blocks: exact=7
                           ->  Bitmap Index Scan on next_attempt_at_partial_idx  (cost=0.00..4.38 rows=32 width=0) (actual time=39.577..39.577 rows=10 loops=2)
                                 Index Cond: (next_attempt_at <= '2024-10-19 22:11:36.310082+00'::timestamp with time zone)
 Planning Time: 3.019 ms
 Execution Time: 7411.459 ms
(31 rows)
```

Out of curiosity, I'll try a composite index and see what difference it makes. Either way, this feels like a big step in the right direction.

Reply:
With a composite index, planning time is 2.5 ms faster, but execution time is 363 ms slower. The partial index wins.

Reply:
I found some changes that made a significant difference. efc4dde3b96803301a2a5ec888b15c38c8941623

The details are in the commit message:

> After some analyzing, it turns out that the query is between two and three times as fast using a partial index on `RepeatRecord.repeater_id` where `state` is Pending or Failed. Performance is also improved using a partial index with `Repeater.is_deleted` instead of a normal B-tree index. When these two indexes are used, the `next_attempt_at_partial_idx` is not used.

I've dropped the `next_attempt_at_partial_idx` because of that, and changed the indexes for `Repeater.state` and `Repeater.is_deleted` to be partial.

Reply:
I think `Parallel Seq Scan on repeaters_repeatrecord` is going to be a problem if that query plan is chosen on prod. In comparison, `Index Scan on repeaters_repeater_...` is inconsequential because there are only ~2000 (tiny number) repeaters whereas there are millions of repeat records.

Reply:
Just to close the loop here: I tested the RepeatRecord.state index on Prod, and Postgres uses it for both subqueries. I tested on Friday afternoon, and the query took around 27 seconds.

Reply:
I think the benefits of this PR are big enough for us to iterate on step-wise improvements after it is merged. One improvement I think we should consider is replacing the `repeaters_repeatrecord` table with nested partitioning / subpartitioning. Something like:

```
CREATE TABLE repeaters_repeatrecord_p (
  -- ...
) PARTITION BY LIST (state);

CREATE TABLE repeaters_repeatrecord_ready
  PARTITION OF repeaters_repeatrecord_p
  FOR VALUES IN (1, 2);

CREATE TABLE repeaters_repeatrecord_done
  PARTITION OF repeaters_repeatrecord_p
  FOR VALUES IN NOT (1, 2)
  PARTITION BY RANGE (registered_on);

CREATE TABLE repeaters_repeatrecord_2010
  PARTITION OF repeaters_repeatrecord_done
  FOR VALUES FROM ('2010-01-01') TO ('2011-01-01');

-- etc.
```

This will improve queries on repeat records ready to be sent, and also allow us to drop completed repeat records after some expiry (TBD).